internal/mesh/internal/controllers/sidecarproxy/controller.go

```diff
@@ -0,0 +1,168 @@
+
+type TrustDomainFetcher func() (string, error)
+
+func Controller(cache *sidecarproxycache.Cache, mapper *sidecarproxymapper.Mapper, trustDomainFetcher TrustDomainFetcher) controller.Controller {
+	if cache == nil || mapper == nil || trustDomainFetcher == nil {
+		panic("cache, mapper and trust domain fetcher are required")
+	}
+
+	return controller.ForType(types.ProxyStateTemplateType).
+		WithWatch(catalog.ServiceEndpointsType, mapper.MapServiceEndpointsToProxyStateTemplate).
+		WithWatch(types.UpstreamsType, mapper.MapDestinationsToProxyStateTemplate).
+		WithReconciler(&reconciler{cache: cache, getTrustDomain: trustDomainFetcher})
+}
+
+type reconciler struct {
+	cache          *sidecarproxycache.Cache
+	getTrustDomain TrustDomainFetcher
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", ControllerName)
+
+	rt.Logger.Trace("reconciling proxy state template", "id", req.ID)
+
+	// Instantiate a data fetcher to fetch all reconciliation data.
+	dataFetcher := fetcher.Fetcher{Client: rt.Client, Cache: r.cache}
+
+	// Check if the workload exists.
+	workloadID := resource.ReplaceType(catalog.WorkloadType, req.ID)
+	workload, err := dataFetcher.FetchWorkload(ctx, resource.ReplaceType(catalog.WorkloadType, req.ID))
+	if err != nil {
+		rt.Logger.Error("error reading the associated workload", "error", err)
+		return err
+	}
+	if workload == nil {
+		// If workload has been deleted, then return as ProxyStateTemplate should be cleaned up
+		// by the garbage collector because of the owner reference.
+		rt.Logger.Trace("workload doesn't exist; skipping reconciliation", "workload", workloadID)
+		return nil
+	}
+
+	proxyStateTemplate, err := dataFetcher.FetchProxyStateTemplate(ctx, req.ID)
+	if err != nil {
+		rt.Logger.Error("error reading proxy state template", "error", err)
+		return nil
+	}
+
+	if proxyStateTemplate == nil {
+		// If proxy state template has been deleted, we will need to generate a new one.
+		rt.Logger.Trace("proxy state template for this workload doesn't yet exist; generating a new one", "id", req.ID)
+	}
+
+	if !fetcher.IsMeshEnabled(workload.Workload.Ports) {
+		// Skip non-mesh workloads.
+
+		// If there's existing proxy state template, delete it.
+		if proxyStateTemplate != nil {
+			rt.Logger.Trace("deleting existing proxy state template because workload is no longer on the mesh", "id", req.ID)
+			_, err = rt.Client.Delete(ctx, &pbresource.DeleteRequest{Id: req.ID})
+			if err != nil {
+				rt.Logger.Error("error deleting existing proxy state template", "error", err)
+				return err
+			}
+
+			// Remove it from cache.
+			r.cache.DeleteSourceProxy(req.ID)
+		}
+		rt.Logger.Trace("skipping proxy state template generation because workload is not on the mesh", "workload", workload.Resource.Id)
+		return nil
+	}
+
+	// First get the trust domain.
+	trustDomain, err := r.getTrustDomain()
```

Comment:
If the trust domain were to change (which i don't think is possible without a big disruption) the reconciler would not notice until it re-reconciled for another reason.

Reply:
Right. I'm assuming trust domain won't change without a restart. do you think that's not the case?

Reply:
It's probably fine to leave it.